docs/extensions/api.md

```diff
@@ -2,425 +2,563 @@ title:      Extensions API
-there are also Blockprocessors which are part of the core BlockParser.
-
-As the parser builds an [ElementTree][ElementTree] object which is later rendered
-as Unicode text, there are also some helpers provided to ease manipulation of
-the tree. Each part of the API is discussed in its respective section below.
-Additionally, reading the source of some [Available Extensions][] may be
-helpful. For example, the [Footnotes][] extension uses most of the features
-documented here.
-
-## Preprocessors {: #preprocessors }
-
-Preprocessors munge the source text before it is passed into the Markdown
-core. This is an excellent place to clean up bad syntax, extract things the
-parser may otherwise choke on and perhaps even store it for later retrieval.
-
-Preprocessors should inherit from `markdown.preprocessors.Preprocessor` and
-implement a `run` method with one argument `lines`. The `run` method of
-each Preprocessor will be passed the entire source text as a list of Unicode
-strings. Each string will contain one line of text. The `run` method should
-return either that list, or an altered list of Unicode strings.
+custom functionality and syntax into the parser. An extension will patch
+into one or more stages of the parser:
 
-A pseudo example:
+* [*Preprocessors*](#preprocessors) alter the source before it is passed to the parser. 
+* [*Block Processors*](#blockprocessors) work with blocks of text separated by blank lines.
+* [*Tree Processors*](#treeprocessors) modify the constructed ElementTree
+* [*Inline Processors*](#inlineprocessors) are common tree processors for inline elements, such as `*strong*`. 
+* [*Postprocessors*](#postprocessors) munge of the output of the parser just before it is returned. 
+
+
+The parser loads text, applies the preprocessors, creates an
+[ElementTree][ElementTree] added to by the block processors and 
+inline patterns, renders the ElementTree as Unicode text, and then 
+then applies the postprocessors.
+
+There are classes and helpers provided to ease writing your extension. 
+Each part of the API is discussed in its respective section below.
+Additionally, you can walk through the [Tutorial on Writing Extensions][tutorial];
+look at some of the [Available Extensions][] and their [source code][extension source].
+As always, you may report bugs, ask for help, and discuss various 
+other issues on the [bug tracker].
+
+## Phases of processing {: #stages }
+### Preprocessors {: #preprocessors }
+
+Preprocessors munge the source text before it is passed to the Markdown
+parser. This is an excellent place to clean up bad characters or to extract 
+portions for later processing that the parser may otherwise choke on.
+
+Preprocessors inherit from `markdown.preprocessors.Preprocessor` and
+implement a `run` method, which takes a single parameter `lines`. This parameter is
+the entire source text stored as a list of Unicode string, one per line.  `run` should
```

Comment:
Missing an `s` here: "... a list of Unicode string<ins>s</ins>, one per line."

Reply:
Thank you.  Done.